Improve exception message generated by `EnhancingX509ExtendedTrustManager` (#16056)

idelpivnitskiy opened the pull request:
Motivation:

This is an improvement for the feature introduced in #13381:
1. Include IP addresses in case matching happens via `HostnameChecker.matchIP`.
2. Include `URI` entries to detect when hostname verification should be disabled.
3. Include common name in case verification fallbacks to `CN` from subject.
4. Include both SNI hostname and peerHost to observe what was used for both paths inside `HostnameChecker`.
5. Split entries by cert in the chain.
6. Catch more use-cases (exception messages) coming from `HostnameChecker`.

Modifications:
- Enhance `EnhancingX509ExtendedTrustManager.throwEnhancedCertificateException` to include DNS, IP, URI types of SAN entries, CN as well as SNI hostname and peerHost.
- Modify tests to assert new behavior.

Result:

More details in the exception message related to full logic of `HostnameChecker`. Example:
```
No subject alternative DNS name matching netty.io found for SNIHostName=netty.io and peerHost=netty.io in the chain of 1 certificate(s): 1. subjectAlternativeNames=[DNS:some.netty.io, IP:127.0.0.1, URI:URI:https://uri.netty.io/profile], CN=leaf.netty.io.
```

idelpivnitskiy committed:
Improve exception message generated by `EnhancingX509ExtendedTrustManager` (#16056)

Motivation:

This is an improvement for the feature introduced in #13381:
1. Include IP addresses in case matching happens via
`HostnameChecker.matchIP`.
2. Include `URI` entries to detect when hostname verification should be
disabled.
3. Include common name in case verification fallbacks to `CN` from
subject.
4. Include both SNI hostname and peerHost to observe what was used for
both paths inside `HostnameChecker`.
5. Split entries by cert in the chain.
6. Catch more use-cases (exception messages) coming from
`HostnameChecker`.

Modifications:
- Enhance
`EnhancingX509ExtendedTrustManager.throwEnhancedCertificateException` to
include DNS, IP, URI types of SAN entries, CN as well as SNI hostname
and peerHost.
- Modify tests to assert new behavior.

Result:

More details in the exception message related to full logic of
`HostnameChecker`. Example:
```
No subject alternative DNS name matching netty.io found for SNIHostName=netty.io and peerHost=netty.io in the chain of 1 certificate(s): 1. subjectAlternativeNames=[DNS:some.netty.io, IP:127.0.0.1, URI:URI:https://uri.netty.io/profile], CN=leaf.netty.io.
```

---------

Co-authored-by: Norman Maurer <norman_maurer@apple.com>
Co-authored-by: Chris Vest <mr.chrisvest@gmail.com>

idelpivnitskiy committed:
Hide Java 8 classes

idelpivnitskiy committed:
Hide more Java 8 classes

chrisvest commented on handler/src/test/java/io/netty/handler/ssl/SniClientJava8TestUtil.java:
```diff
@@ -345,4 +347,11 @@ public String chooseEngineServerAlias(String s, Principal[] principals,
+
+    static SSLSession mockSSLSessionWithSNIHostNameAndPeerHost(String hostname) {
+        ExtendedSSLSession session = Mockito.mock(ExtendedSSLSession.class);
+        Mockito.when(session.getRequestedServerNames()).thenReturn(Arrays.asList(new SNIHostName(hostname)));
```
```suggestion
        SNIServerName sniName = new SNIHostName(hostname);
        Mockito.when(session.getRequestedServerNames()).thenReturn(Arrays.asList(sniName));
```

chrisvest reviewed (commented):
We need to massage the types a bit to make the compiler happy.

chrisvest committed:
Fix compilation error

chrisvest committed:
Fix class verifier error from using static method calls on interfaces

Static method interface calls are not supported in Java 6 byte code, which we target.

chrisvest commented:
Thanks!
